Use before declaration false positive when combined with binding exports

**TypeScript Version:**  2.2.1

**Code**

```ts
import methodFactory from './http-client'

export { httpDelete as delete } // block scoped httpDelete used before declaration
const httpDelete = methodFactory('DELETE')
```

**Expected behavior:**

No error. The export declaration makes no use of the variable; what should happen is that `httpDelete`'s TDZ should extend to the `delete` export, but that happens regardless of the relative position of the `export` declaration.

**Actual behavior:**

Two errors in the same line about the variable being used before assignment.

This error is correct in the case of `export default` and the typescript-specific `export =`, but not in any other use of `export`.

Does this apply to classes as well?

With the current nightly (`typescript@2.3.0-dev.20170403`) I get an error for exporting classes before their declaration.

```ts
export { XYZ }
//       ~~~
// XYZ.ts(1,10): error TS2449: Class 'XYZ' used before its declaration.

class XYZ {
}
```

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -742,6 +742,13 @@ namespace ts {
             // declaration is after usage, but it can still be legal if usage is deferred:
             // 1. inside a function
             // 2. inside an instance property initializer, a reference to a non-instance property
+            // 3. inside an export specifier
+
+            if (usage.parent.kind === SyntaxKind.ExportSpecifier) {
+                // export specifiers do not use the variable, they only make it available for use
+                return true;
+            }
+
             const container = getEnclosingBlockScopeContainer(declaration);
             const isInstanceProperty = declaration.kind === SyntaxKind.PropertyDeclaration && !(getModifierFlags(declaration) & ModifierFlags.Static);
             return isUsedInFunctionOrInstanceProperty(usage, isInstanceProperty, container);

--- a/tests/baselines/reference/exportBinding.js
+++ b/tests/baselines/reference/exportBinding.js
@@ -1,0 +1,19 @@
+//// [exportBinding.ts]
+export { x }
+const x = 'x'
+
+export { Y as Z }
+class Y {}
+
+
+//// [exportBinding.js]
+"use strict";
+exports.__esModule = true;
+var x = 'x';
+exports.x = x;
+var Y = (function () {
+    function Y() {
+    }
+    return Y;
+}());
+exports.Z = Y;

--- a/tests/baselines/reference/exportBinding.symbols
+++ b/tests/baselines/reference/exportBinding.symbols
@@ -1,0 +1,14 @@
+=== tests/cases/conformance/es6/modules/exportBinding.ts ===
+export { x }
+>x : Symbol(x, Decl(exportBinding.ts, 0, 8))
+
+const x = 'x'
+>x : Symbol(x, Decl(exportBinding.ts, 1, 5))
+
+export { Y as Z }
+>Y : Symbol(Z, Decl(exportBinding.ts, 3, 8))
+>Z : Symbol(Z, Decl(exportBinding.ts, 3, 8))
+
+class Y {}
+>Y : Symbol(Y, Decl(exportBinding.ts, 3, 17))
+

--- a/tests/baselines/reference/exportBinding.types
+++ b/tests/baselines/reference/exportBinding.types
@@ -1,0 +1,15 @@
+=== tests/cases/conformance/es6/modules/exportBinding.ts ===
+export { x }
+>x : "x"
+
+const x = 'x'
+>x : "x"
+>'x' : "x"
+
+export { Y as Z }
+>Y : typeof Y
+>Z : typeof Y
+
+class Y {}
+>Y : Y
+

--- a/tests/cases/conformance/es6/modules/exportBinding.ts
+++ b/tests/cases/conformance/es6/modules/exportBinding.ts
@@ -1,0 +1,5 @@
+export { x }
+const x = 'x'
+
+export { Y as Z }
+class Y {}

